### PR TITLE
Added TTL query field to features API.

### PR DESCRIPTION
**Pull request by elland**
## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (elland) [inline libs/wire-api/src/Wire/API/Team/Feature.hs:281]: Do we have any natural implementation in use? Is it worth going for more type-level constriction here or documenting externally that `TTL=0` is ignored?

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (elland) [inline libs/wire-api/src/Wire/API/Team/Feature.hs:299]: Errors are currently not surfaced, everything turns into a generic 500. Maybe we should address that. :)

File: services/galley/src/Galley/Cassandra/TeamFeatures.hs
Comment (elland) [inline services/galley/src/Galley/Cassandra/TeamFeatures.hs:106]: Do we already have a way to handle this kind of stuff in the codebase?

**Review by elland — COMMENTED** [review]
Left some open questions.

File: services/galley/src/Galley/Cassandra/TeamFeatures.hs
Comment (battermann) [inline services/galley/src/Galley/Cassandra/TeamFeatures.hs:108]: These 3 function read nicely, at the same time it's not so clear what they really do. `days` converts days to seconds, right? What about `fromDays` or `toSeconds`?

File: services/galley/src/Galley/API/Teams/Features.hs
Comment (battermann) [inline services/galley/src/Galley/API/Teams/Features.hs:572]: Maybe it's not very important, but something to consider, when the TTL expires, and the feature is disabled again, this will not trigger a push event. We should check if this can cause any problems, when clients do not get this info via push event right away.

**Review by battermann — COMMENTED** [review]
Looking good
An integration would be nice
Does it make sense to use days for the TTL in stern and seconds in the galley API? In this case we can have an integration test with a very short TTL.

File: services/galley/src/Galley/Cassandra/TeamFeatures.hs
Comment (elland) [reply services/galley/src/Galley/Cassandra/TeamFeatures.hs:108]: Yeah, the naming is not great, I just copied it over from https://hackage.haskell.org/package/timespan-0.4.0.0/docs/src/Data.Time.TimeSpan.html#days :D

File: services/galley/src/Galley/API/Teams/Features.hs
Comment (elland) [reply services/galley/src/Galley/API/Teams/Features.hs:572]: For this specific feature or any feature?

File: services/galley/src/Galley/Cassandra/TeamFeatures.hs
Comment (battermann) [inline services/galley/src/Galley/Cassandra/TeamFeatures.hs:116]:
```suggestion
            _ -> "USING TTL 0"
```

File: services/galley/src/Galley/Cassandra/TeamFeatures.hs
Comment (battermann) [reply services/galley/src/Galley/Cassandra/TeamFeatures.hs:106]: I am not aware

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (fisx) [reply libs/wire-api/src/Wire/API/Team/Feature.hs:302]: but this `e` should result in a 400 generated by servant, no?

in general: yes, if you find an easy way to pass on error objects from galley intact to the stern "UI", please do so.  if it's hard for any reason, please open a ticket.

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Feature.hs]: Use `readEither` here, too, and `fail` if `Left`?

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Feature.hs:337]: this is old swagger-1.2 code, but it looks like we still need this?

File: nix/default.nix
Comment (fisx) [inline nix/default.nix:57]: separate PR?  it's ok if you're lazy and somebody approves, just couldn't keep myself from raising an eyebrow at least.  :)

File: services/galley/src/Galley/API/Teams/Features.hs
Comment (fisx) [inline services/galley/src/Galley/API/Teams/Features.hs:631]: i like `_ttl` slightly better than `_`.  :)

File: services/galley/src/Galley/Cassandra/TeamFeatures.hs
Comment (fisx) [reply services/galley/src/Galley/Cassandra/TeamFeatures.hs]: it's local names, so names don't need to be awfully long.

File: services/galley/src/Galley/Cassandra/TeamFeatures.hs
Comment (fisx) [reply services/galley/src/Galley/Cassandra/TeamFeatures.hs]: i found https://github.com/wireapp/wire-server/blob/4c2dc58c75ce25da228f6e9dc1069a82e628fc26/services/brig/src/Brig/Unique.hs#L68, but it's not used exactly exhaustively...

File: services/galley/src/Galley/API/Internal.hs
Comment (fisx) [inline services/galley/src/Galley/API/Internal.hs:438]: what's this about?

**Review by fisx — COMMENTED** [review]
looks good so far!

File: services/galley/src/Galley/API/Internal.hs
Comment (fisx) [reply services/galley/src/Galley/API/Internal.hs:438]: ah, now i see the difference between the two definitions.  :)  if you don't want to add a comment, maybe it'll help future me if you change `fsSetD` to `fsSetDeprecated`?

File: changelog.d/1-api-changes/added-ttl-param-to-back-office-feature-api
Comment (fisx) [inline changelog.d/1-api-changes/added-ttl-param-to-back-office-feature-api:1]:
```suggestion
The back-office (aka stern) team feature API now accenpts an optional TTL parameter (in days), so features can be activated for a limited period.
```

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Feature.hs:278]: do you even need these?  (just wondering, i don't mind them being there just in case.)

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Feature.hs:298]: I sometimes like to implement `FromHttpApiData` in terms of `FromByteString` or vice versa.  Makes for more coherent error handling, for once.  You could also factor out a function that is called in both instances.  But I don't insist.

**Review by fisx — COMMENTED** [review]
please sync with the people working on https://github.com/wireapp/wire-server/pull/2435 before merging, i'm sure there will be lots of ugly conflicts...

File: libs/wire-api/src/Wire/API/Routes/Public/Galley.hs
Comment (fisx) [inline libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:1179]: no time to check, but is this optional (with default "unlimited"?  if not please change it, to make sure ibis won't break over this change!

Comment (elland) [reply libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:1179]: Should the default be `unlimited` or 8 weeks?

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (elland) [reply libs/wire-api/src/Wire/API/Team/Feature.hs:275]: `ToBytestring` is used to serialise it for Bilge. `ToHttpApiData` on the other hand, doesn't seem necessary. 👍

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (elland) [reply libs/wire-api/src/Wire/API/Team/Feature.hs:295]: Removing one of them anyway, so no need to 🏌️‍♂️it. :)

File: services/galley/test/integration/API/Teams/Feature.hs
Comment (fisx) [inline services/galley/test/integration/API/Teams/Feature.hs:389]: uncomment

File: tools/stern/src/Stern/API.hs
Comment (elland) [reply tools/stern/src/Stern/API.hs:383]: Good catch! Would be nice to have tests through Stern.

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Feature.hs]: I remember that we don't want this extension, but I don't remember why.  I might be wrong.

But the json instances created weren't very nice:

```
*Wire.API.Team.Feature Data.Aeson> encode TeamFeatureTTLUnlimited
"{\"tag\":\"TeamFeatureTTLUnlimited\"}"
*Wire.API.Team.Feature Data.Aeson> encode (TeamFeatureTTLSeconds 3)
"{\"contents\":3,\"tag\":\"TeamFeatureTTLSeconds\"}"
```

Which is a nice case for adding golden tests for new types as well as keeping the old ones, come to think of it.  :)

File: libs/wire-api/src/Wire/API/Team/Feature.hs
Comment (fisx) [inline libs/wire-api/src/Wire/API/Team/Feature.hs:301]: would this have been really bad (unlimited enablement would have been disabled immediately again), or no problem (0 means "indefinite" in cassandra as well)?  either way, it should be fixed, and we should have roundtrip tests for this, and there should be an error if a cql instance does not have a test, and...

File: tools/stern/src/Stern/Intra.hs
Comment (fisx) [inline tools/stern/src/Stern/Intra.hs:575]: This still looks wrong.  We're parsing something that we call seconds, and then we interpret it as days.  Introduce a type `TeamFeatureTTLDays` here?  I know it's annoying, so I'll leave it to you to decide whether it's worth it.

**Review by fisx — APPROVED** [review]
Found a few more details.  Please take a look at my changes and comments.  No further review needed if you're happy.

File: tools/stern/src/Stern/Intra.hs
Comment (elland) [reply tools/stern/src/Stern/Intra.hs:575]: I went back and forth on this a few times but never got to a point I was happy with. We end up with either a "mirror" type, that does Unlimited/Days and converts to Unlimited/Seconds, or an extra, used only once constructor for Days that we can't disambiguate on parsing when receiving it via the API call.

I'm open to suggestions, but maybe this would be easier if/when we servantify stern?

Comment (fisx) [reply tools/stern/src/Stern/Intra.hs:575]: maybe the solution would be something like this:

```haskell
data TTL = TTLSecs Int | TTLUnlimited

parse "3s" = TTLSecs 3
parse "1d" = TTLSecs $ 1 * 60 * 60 * 24
...
```

anyway, yes, let's keep it as is for now.

**Review by fisx — APPROVED** [review]

**Pull request merged**
